### ISSUE TITLE
How can I build/update the patch myself for the newest version?

### ISSUE THREAD

**Issue by odomingao** [anthraxx/linux-hardened#96]
Hi,

I'd like to update the patch myself for the latest version of the kernel. How can I do this?

(edit: if this is something that is simple/can be automated, that is)

Thanks.

**Comment by hardfalcon** [comment]
This can be done by rebasing the patches from a certain branch.
Although *in principal*, the process is always the same, I want to distinguish between two basic scenarios:

1. You might be **lucky** and the only patch that requires manual intervention/fixing during the rebase is the very last patch of the whole patchset (the one that sets `EXTRAVERSION` to `-hardened1` in the `Makefile`). In most (but not all) cases, this scenario is what happens when you try a rebase within the same version branch. One example for this scenario would be rebasing the hardened patchset from `v6.10.11-hardened1` onto vanilla kernel `v6.10.12`.
2. You might be **unlucky** and have to manually fix/rebase more than just the above mentioned `-hardened1` patch.
   Sometimes, this scenario occurs even for rebases within the same version branch. For example, if you try to rebase tag `v6.10.12-hardened1` onto the last 6.10 kernel tag `v6.10.14`, you will run into a few additional merge conflicts that require manual fixes during the rebase process.
   Also, this scenario occurs almost always when you try to rebase the hardened patchset from one version branch onto another version branch (for example, when you try to rebase the hardened patchset from the `v6.10.12-hardened1` tag onto vanilla kernel `v6.11.3`). Although this is *in principal* still the same problem, in *practice*, rebases across different version branches will usually require more manual fixes than a rebase within the same version branch, and some of those manual fixes will oftentimes be more difficult to achieve and demand more expertise.
   
Important: Be aware that a (seemingly) clean `git rebase` is *not* a guarantee that the resulting patchset will work as expected. Thankfully, this doesn't happen very often, but *sometimes*, rebases *can* introduce bugs even though they went through seemingly smoothly without requiring manual fixes. Though a rare occurrence, it's important to be aware that this is a possibility and can cause compile errors (if you're lucky) or (if you are unlucky) even nasty runtime bugs.

Here's a "universal" howto that describes how to rebase regardless of what the current/latest version in any of the version branches actually is.

Only perform these steps the first time:
```bash
# Clone this repo (this will take a while)
git clone https://github.com/anthraxx/linux-hardened.git

# Change into your local clone of the repo(s)
cd linux-hardened

# Rename the remote from "origin" to an easier to recognize name
# (not strictly necessary, but makes this howto easier to understand)
git remote rename origin remote_linux-hardened

# Add remote for the vanilla stable kernel
git remote add remote_linux-stable https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git
```

Execute these steps every time you want to rebase the hardened patchset:
```bash
# Change into your local clone of the repos
cd linux-hardened

# Fetch all available updates from all remotes, and explicitly include all tags as well
git fetch --all --tags

# Check if your tree is clean (no uncommitted changes, no untracked files, no unfinished rebase underway, etc)
git status

# Only if your tree is *not* clean, you may have to do one or more of these commands:
git rebase --about
git reset --hard

# Start the actual rebase.
# 
# In this example, we rebase all changes/commits between the v6.10.11 tag and
# the v6.10.11-hardened1 tag onto the v6.10.12 tag.
# I've chosen these versions/tags for two reasons:
# 1) They lead to the "lucky" scneario described above, making this howto easier to follow.
# 2) There is already an "official" linux-hardened-v6.10.12-hardened1.patch, which allows you to compare results.
# 
# The "--interactive" parameter is optional but recommended because
# it shows you the list of commits that  you are actually rebasing.
# 
# Note: It is *not* important that there is only one "version step" between "6.10.11" and "6.10.12".
# 
# However, it *is* important/recommended that the two tags defining
# the begin and the end of the list of patches to be rebased (here: "v6.10.11" and "v6.10.11-hardened1")
# are the same kernel version, so you only rebase the actual linux-hardened patchset and no other patches.
# 
git rebase --interactive --onto v6.10.12 v6.10.11 v6.10.11-hardened1

# After running the above command, an editor will open with a list of the commits to be rebased.
# In our example, there should be 103 commits on the list, and the first entry in the list should be:
# pick 1e21d1486bfe make DEFAULT_MMAP_MIN_ADDR match LSM_MMAP_MIN_ADDR
# The last entry in the list should be:
# pick ebda5a529308 Linux hardened v6.10.11-hardened1
# Exit that edit (for example, using Esc :q in vim) to proceed with the rebase.

# You should now see git iterating through the 103 patches.
# After a few seconds, you should get the following error message:
# 
# Auto-merging Makefile
# CONFLICT (content): Merge conflict in Makefile
# error: could not apply ebda5a529308... Linux hardened v6.10.11-hardened1
# hint: Resolve all conflicts manually, mark them as resolved with
# hint: "git add/rm <conflicted_files>", then run "git rebase --continue".
# hint: You can instead skip this commit: run "git rebase --skip".
# hint: To abort and get back to the state before "git rebase", run "git rebase --abort".
# hint: Disable this message with "git config advice.mergeConflict false"
# Could not apply ebda5a529308... Linux hardened v6.10.11-hardened1
#
# Note: Even though in this case, the merge conflict is only a single change ("patch hunk") in a single file,
# merge conflicts can occur in multiple files at the same tame, and with multiple patch hunks within the same file.
# 
# Edit the file "./Makefile" and resolve the conflict.
# Consider changing the value for "EXTRAVERSION" to a different string,
# perhaps "-hardened0" instead of "-hardened1".
# Also, remove the marker lines between the two conflicting hunks/sections, then save the file.
vim ./Makefile

# Add the changed file to the commit we are currently rebasing:
git add ./Makefile

# Continue the rebase.
# An editor with the commit message will open.
# In most *other* cases, we normally would *not* change the commit message.
# However, in this specific case, you should probably change the kernel version and the "Signed-off-by:" line.
# Also, consider changing "-hardened1" into "-hardened0", if you have done the same with the Makefile.
# When you are finished with editing the commit message, save the file and exit the editor.
git rebase --continue

# Create a tag so you don't lose your work when checking out other branches/tags.
# Important: Always use a naming/tagging scheme that does not collide with the tag names from the remotes you are using.
# For example, you could simply append your nickname to the tags that you create.
git tag v6.10.12-hardened0-odomingao

# Finally, create your very own "linux-hardened-v6.10.12-hardened0-odomingao.patch" file:
git diff --output=../linux-hardened-v6.10.12-hardened0-odomingao.patch v6.10.12 v6.10.12-hardened0-odomingao
```

Additional remarks:
- Theoretically, you can also use commit IDs and/or branch names with the `git rebase` command. However, especially using branch names can make your life more difficult because it can "pollute" your local branch in a way that makes future `git pull`s from the remote/upstream branches unnecessary difficult or even downright impossible.
- When the first release of a new kernel version branch is released (for example, `v6.11`), that tag will usually only appear in the `linux-stable` repo after several days. If you want that tag ASAP, or if you want to play around with release candidates of the next version branch, you can add Linus Torvalds' remote:
  ```bash
  git remote add remote_linux-torvalds https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git
  ```
- If you want to rebase onto a release candidate of the next minor release from one of the stable kernel version branches, there's also a remote for that:
  ```bash
  git remote add remote_linux-stable-rc https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux-stable-rc.git
  ```

**Comment by odomingao** [comment]
Thank you very much for the detailed response!
